source/endpoints/RuleEngine.cpp

```diff
@@ -0,0 +1,177 @@
+            }
+
+            ResolutionOutcome::ResolutionOutcome(aws_endpoints_resolved_endpoint *impl, Allocator *allocator)
+                : m_allocator(allocator), m_resolvedEndpoint(impl)
+            {
+            }
+
+            ResolutionOutcome::ResolutionOutcome(ResolutionOutcome &&toMove) noexcept
+                : m_resolvedEndpoint(toMove.m_resolvedEndpoint)
+            {
+                toMove.m_resolvedEndpoint = nullptr;
+            }
+
+            ResolutionOutcome &ResolutionOutcome::operator=(ResolutionOutcome &&toMove)
+            {
+                if (&toMove != this)
+                {
+                    *this = ResolutionOutcome(std::move(toMove));
+                }
+
+                return *this;
+            }
+
+            ResolutionOutcome::~ResolutionOutcome()
+            {
+                aws_endpoints_resolved_endpoint_release(m_resolvedEndpoint);
+            }
+
+            bool ResolutionOutcome::IsEndpoint() const noexcept
+            {
+                return AWS_ENDPOINTS_RESOLVED_ENDPOINT == aws_endpoints_resolved_endpoint_get_type(m_resolvedEndpoint);
+            }
+
+            bool ResolutionOutcome::IsError() const noexcept
+            {
+                return AWS_ENDPOINTS_RESOLVED_ERROR == aws_endpoints_resolved_endpoint_get_type(m_resolvedEndpoint);
+            }
+
+            Optional<StringView> ResolutionOutcome::GetUrl() const noexcept
+            {
+                ByteCursor url;
+                if (aws_endpoints_resolved_endpoint_get_url(m_resolvedEndpoint, &url))
+                {
+                    return Optional<StringView>();
+                }
+
+                return Optional<StringView>(ByteCursorToStringView(url));
+            }
+
+            inline StringView CrtStringToStringView(const aws_string *s)
+            {
+                ByteCursor key = aws_byte_cursor_from_string(s);
+                return ByteCursorToStringView(key);
+            }
+
+            Optional<UnorderedMap<StringView, Vector<StringView>>> ResolutionOutcome::GetHeaders() const noexcept
+            {
+                const aws_hash_table *resolved_headers = nullptr;
+
+                if (aws_endpoints_resolved_endpoint_get_headers(m_resolvedEndpoint, &resolved_headers))
+                {
+                    return Optional<UnorderedMap<StringView, Vector<StringView>>>();
+                }
+
+                UnorderedMap<StringView, Vector<StringView>> headers;
+                for (struct aws_hash_iter iter = aws_hash_iter_begin(resolved_headers); !aws_hash_iter_done(&iter);
+                     aws_hash_iter_next(&iter))
+                {
+                    ByteCursor key = aws_byte_cursor_from_string((const aws_string *)iter.element.key);
+                    const aws_array_list *array = (const aws_array_list *)iter.element.value;
+                    headers.insert(
```

Comment:
although `insert` may take the most efficient copy-elision overload,
have you tried/considered `emplace` to be sure that copy of a temporary object does not happen?

Reply:
yeah, emplace would be better here.